v1/common/amqp.go

```diff
@@ -4,23 +4,47 @@ import (
 	"crypto/tls"
 	"fmt"
 	"sync"
+	"time"
 
 	"github.com/streadway/amqp"
 )
 
 // AMQPConnector ...
 type AMQPConnector struct {
 	connManager *amqpConnectionManager
+
+	exchangeMaxRetries   int
+	exchangeRetryTimeout time.Duration
 }
 
 func NewAMQPConnector(url string, tlsConfig *tls.Config) *AMQPConnector {
 	return &AMQPConnector{
 		connManager: newAMQPConnectionManager(url, tlsConfig),
+
+		exchangeMaxRetries:   3,
+		exchangeRetryTimeout: time.Second,
 	}
 }
 
 // Exchange declares an exchange, opens a channel declares and binds the queue and enables publish notifications using the existing RabbitMQ connection.
 func (ac *AMQPConnector) Exchange(exchange, exchangeType, queueName string, queueDurable, queueDelete bool, queueBindingKey string, exchangeDeclareArgs, queueDeclareArgs, queueBindingArgs amqp.Table) (*amqp.Channel, amqp.Queue, <-chan amqp.Confirmation, error) {
+	var lastErr error
+	for retry := 0; retry < ac.exchangeMaxRetries; retry++ {
+		channel, queue, confirmChan, err := ac.exchange(exchange, exchangeType, queueName, queueDurable, queueDelete, queueBindingKey, exchangeDeclareArgs, queueDeclareArgs, queueBindingArgs)
+		if err != nil {
+			lastErr = err
+			_, connErr := err.(amqpConnError)
+			if connErr || err == amqp.ErrClosed {
```

Comment:
You should trigger a refresh of the connection here, not just retry otherwise it will fail again.

Reply:
the next call will call `ac.connManager.get()`, which:
- if the connection is `nil`, creates a new one
- if the connection closed in the meantime, creates a new one
- if the connection attempt timed out previously, creates a new one (it's `nil`)